### PR TITLE
New feature - a concurrency limiter

### PR DESCRIPTION
**Pull request by spkrka**

### CODE REVIEW

File: src/test/java/com/spotify/futures/jdk7/ConcurrencyLimiterTest.java [hunk 1]
```diff
@@ -0,0 +1,241 @@
...
+    @Override
+    public ListenableFuture<String> call() throws Exception {
+      activeCount.incrementAndGet();
+      if (activeCount.get() > maxCount.get()) {
+        maxCount.set(activeCount.get());
+      }
+      return future;
+    }
+  }
+}
```
Comment (odsod) [inline src/test/java/com/spotify/futures/jdk7/ConcurrencyLimiterTest.java]: Consider adding missing newline.

File: README.md [hunk 2]
```diff
@@ -177,6 +177,24 @@ final ListenableFuture<B> future = getFuture();
...
+To help with this, there is a new class called `ConcurrencyLimiter`. You use it like this:
+
+```java
+final int maxConcurrency = 10;
+final ConcurrencyLimiter<T> = ConcurrencyLimiter.create(maxConcurrency);
+for (int i = 0; i < 1000; i++) {
+  final ListenableFuture<T> future = concurrencyLimiter.add(() -> createFuture());
+}
+```
+The concurrency limiter will ensure that at most 10 futures are created and incomplete at the same time. All the jobs that
```
Comment (odsod) [inline README.md]: Line length seems a bit random, max 80 chars makes the raw text file more readable, I think.

**Comment by odsod** [comment]
+1

Very nice, this will come in handy.

File: src/test/java/com/spotify/futures/jdk7/ConcurrencyLimiterTest.java (hunk 1 above)
Comment (spkrka) [inline src/test/java/com/spotify/futures/jdk7/ConcurrencyLimiterTest.java]: done

File: README.md (hunk 2 above)
Comment (spkrka) [inline README.md]: done

**Comment by pettermahlen** [comment]
Looks good, a couple of questions, though:

- did you consider (optionally) using a BlockingQueue, so as to enforce size limits? The add() method of a BlockingQueue will throw when trying to add beyond a certain limit.
- the size() method of ConcurrentLinkedQueue is pretty slow, according to its javadocs, so calling numQueued() may be expensive. Should that be documented somehow?

**Comment by spkrka** [comment]
@pettermahlen Thanks comments.

* Creating a limiter requires specifying a maximum queue size. Unbounded queues are dangerous after all.
* Calling add may throw an exception if the maximum queue size is exceeded.
* Instead of calling queue.size(), the code tracks the size manually with an AtomicInteger.

File: src/main/java/com/spotify/futures/ConcurrencyLimiter.java [hunk 3]
```diff
@@ -0,0 +1,177 @@
...
+   *          input function.
+   * @throws {@link NullPointerException} if callable is null
+   * @throws {@link CapacityReachedException} when soft queue size limit is exceeded.
+   */
+  public ListenableFuture<T> add(Callable<? extends ListenableFuture<T>> callable) {
+    Preconditions.checkNotNull(callable);
+    final SettableFuture<T> response = SettableFuture.create();
+    final Job<T> job = new Job<T>(callable, response);
+    if (queueSize.get() >= maxQueueSize) {
+      throw new CapacityReachedException("Queue size has reached capacity: " + maxQueueSize);
```
Comment (rschildmeijer) [inline src/main/java/com/spotify/futures/ConcurrencyLimiter.java]: I don't like mixing sync exc with async exc.

**Comment by rouzwawi** [comment]
I'm a bit unsure if the calls to `pump()` from the future callback in `invoke()` is such a good idea. It will cause the callables in the queue to be invoked from arbitrary other thread pools in the application usin this utility.

It might be OK from a couple utilization view since the callables should return immediatly, but can cause confusion for other surprising reasons.

**Comment by spkrka** [comment]
Yes, callables will indeed run on seemingly random threads. I don't think this is enough of a problem to block releasing the feature. If it turns out to be a real problem we can create a new API that lets people pass in an executor to run on.

**Pull request merged**